### PR TITLE
Add minibatch

### PR DESCRIPTION
**Pull request by Jerry-Jzy**
##what is the purpose of this pull request?
add minibatch

### CODE REVIEW

File: deepxde/model.py [hunk 1]
```diff
@@ -521,6 +521,7 @@ def train(
         self,
         iterations=None,
         batch_size=None,
+        minibatch=None,
```
Comment (lululxvi) [inline deepxde/model.py:524]: Don't add `minibatch`. `batch_size` is a tuple if there are two batch sizes.

Comment (Jerry-Jzy) [reply deepxde/model.py]: fixed

**Comment by lululxvi** [comment]
Update the docstring.

File: deepxde/data/triple.py [hunk 2]
```diff
@@ -72,13 +72,21 @@ def __init__(self, X_train, y_train, X_test, y_test):
         self.test_x, self.test_y = X_test, y_test
 
         self.train_sampler = BatchSampler(len(X_train[0]), shuffle=True)
```
Comment (lululxvi) [inline deepxde/data/triple.py:74]: `self.train_sampler` --> `self.branch_sampler`

File: deepxde/data/triple.py [hunk 3]
```diff
@@ -72,13 +72,21 @@ def __init__(self, X_train, y_train, X_test, y_test):
...
+        self.trunk_sampler = BatchSampler(len(X_train[1]), shuffle=True)
 
     def losses(self, targets, outputs, loss_fn, inputs, model, aux=None):
         return loss_fn(targets, outputs)
 
     def train_next_batch(self, batch_size=None):
         if batch_size is None:
             return self.train_x, self.train_y
+        if len(batch_size) == 2:
```
Comment (lululxvi) [inline deepxde/data/triple.py:83]: Bug here. What is batch_size is a scaler?

Comment (Jerry-Jzy) [reply deepxde/data/triple.py]: fixed

File: deepxde/data/triple.py (hunk 2 above)
Comment (Jerry-Jzy) [reply deepxde/data/triple.py]: fixed

File: deepxde/model.py [hunk 4]
```diff
@@ -533,10 +533,11 @@ def train(
         Args:
             iterations (Integer): Number of iterations to train the model, i.e., number
                 of times the network weights are updated.
-            batch_size: Integer or ``None``. If you solve PDEs via ``dde.data.PDE`` or
-                ``dde.data.TimePDE``, do not use `batch_size`, and instead use
-                `dde.callbacks.PDEResidualResampler
-                <https://deepxde.readthedocs.io/en/latest/modules/deepxde.html#deepxde.callbacks.PDEResidualResampler>`_,
+            batch_size: Integer, turple, list or ``None``. if you want to use minibatch for trunk net input,
+            set `batch_size` as a turple or list,
```
Comment (lululxvi) [inline deepxde/model.py:537]:
```
batch_size: Integer, tuple, or ``None``. ...
```
Wrong format. https://deepxde--977.org.readthedocs.build/en/977/modules/deepxde.html#deepxde.model.Model.train

File: deepxde/data/triple.py [hunk 5]
```diff
@@ -71,15 +71,23 @@ def __init__(self, X_train, y_train, X_test, y_test):
...
+        self.trunk_sampler = BatchSampler(len(X_train[1]), shuffle=True)
 
     def losses(self, targets, outputs, loss_fn, inputs, model, aux=None):
         return loss_fn(targets, outputs)
 
     def train_next_batch(self, batch_size=None):
         if batch_size is None:
             return self.train_x, self.train_y
-        indices = self.train_sampler.get_next(batch_size)
+        if isinstance(batch_size, (tuple,list)):
```
Comment (lululxvi) [inline deepxde/data/triple.py:83]:
```
if not isinstance(batch_size, (tuple,list)):
    ...
...
```
Also, reformat via black.

Comment (Jerry-Jzy) [reply deepxde/data/triple.py]: fixed

File: deepxde/model.py (hunk 4 above)
Comment (Jerry-Jzy) [reply deepxde/model.py]: fixed

File: deepxde/model.py [hunk 6]
```diff
@@ -533,8 +533,9 @@ def train(
         Args:
             iterations (Integer): Number of iterations to train the model, i.e., number
                 of times the network weights are updated.
-            batch_size: Integer or ``None``. If you solve PDEs via ``dde.data.PDE`` or
-                ``dde.data.TimePDE``, do not use `batch_size`, and instead use
+            batch_size: Integer, tuple, or ``None``. if you want to use minibatch for trunk net input,
```
Comment (lululxvi) [inline deepxde/model.py:536]:
```
batch_size: Integer, tuple, or ``None``.

    - If you solve PDEs via ``dde.data.PDE`` or ``dde.data.TimePDE``, ...
    - For DeepONet, if you want to use mini-batch for trunk net input, ...
```

File: deepxde/data/triple.py [hunk 7]
```diff
@@ -71,16 +71,24 @@ def __init__(self, X_train, y_train, X_test, y_test):
...
     def train_next_batch(self, batch_size=None):
         if batch_size is None:
             return self.train_x, self.train_y
-        indices = self.train_sampler.get_next(batch_size)
-        return (self.train_x[0][indices], self.train_x[1]), self.train_y[indices]
+        if not isinstance(batch_size, (tuple, list)):
+            indices = self.branch_sampler.get_next(batch_size)
+            return (self.train_x[0][indices], self.train_x[1]), self.train_y[indices]
+        indices = self.branch_sampler.get_next(batch_size[0])
+        minibatch_indices = self.trunk_sampler.get_next(batch_size[1])
```
Comment (lululxvi) [inline deepxde/data/triple.py:87]: `indices` --> `indices_branch`
`minibatch_indices` --> `indices_trunk`

File: deepxde/model.py (hunk 6 above)
Comment (Jerry-Jzy) [reply deepxde/model.py]: fixed

File: deepxde/data/triple.py (hunk 7 above)
Comment (Jerry-Jzy) [reply deepxde/data/triple.py]: fixed

File: deepxde/model.py [hunk 8]
```diff
@@ -533,11 +533,13 @@ def train(
         Args:
             iterations (Integer): Number of iterations to train the model, i.e., number
                 of times the network weights are updated.
-            batch_size: Integer or ``None``. If you solve PDEs via ``dde.data.PDE`` or
-                ``dde.data.TimePDE``, do not use `batch_size`, and instead use
-                `dde.callbacks.PDEResidualResampler
-                <https://deepxde.readthedocs.io/en/latest/modules/deepxde.html#deepxde.callbacks.PDEResidualResampler>`_,
-                see an `example <https://github.com/lululxvi/deepxde/blob/master/examples/diffusion_1d_resample.py>`_.
+            batch_size: Integer, tuple, or ``None``.
```
Comment (lululxvi) [inline deepxde/model.py:536]: Format not correct, see https://deepxde--977.org.readthedocs.build/en/977/modules/deepxde.html#deepxde.model.Model.train

Add an empty line before 537

File: deepxde/model.py [hunk 9]
```diff
@@ -533,11 +533,13 @@ def train(
...
+                - If you solve PDEs via ``dde.data.PDE`` or``dde.data.TimePDE``,do not use `batch_size`,and instead use
+                    `dde.callbacks.PDEResidualResampler
+                    <https://deepxde.readthedocs.io/en/latest/modules/deepxde.html#deepxde.callbacks.PDEResidualResampler>`_,
+                    see an `example <https://github.com/lululxvi/deepxde/blob/master/examples/diffusion_1d_resample.py>`_.
+                - For DeepONet, if you want to use mini-batch for trunk net input,
+                    set `batch_size` as a tuple, fist location is the batch size and second location is mini-batch size.
```
Comment (lululxvi) [inline deepxde/model.py:542]:
```
For DeepONet in the format of Cartesian product, if `batch_size` is an Integer, then it is the batch size for the branch input; if you want to also use mini-batch for the trunk net input, set `batch_size` as a tuple, where the fist number is the batch size for the branch net input and the second number is the batch size for the trunk net input.
```

File: deepxde/model.py [hunk 10]
```diff
@@ -533,11 +533,13 @@ def train(
         Args:
             iterations (Integer): Number of iterations to train the model, i.e., number
                 of times the network weights are updated.
-            batch_size: Integer or ``None``. If you solve PDEs via ``dde.data.PDE`` or
-                ``dde.data.TimePDE``, do not use `batch_size`, and instead use
-                `dde.callbacks.PDEResidualResampler
-                <https://deepxde.readthedocs.io/en/latest/modules/deepxde.html#deepxde.callbacks.PDEResidualResampler>`_,
-                see an `example <https://github.com/lululxvi/deepxde/blob/master/examples/diffusion_1d_resample.py>`_.
+            batch_size: Integer, tuple, or ``None``.
+                - If you solve PDEs via ``dde.data.PDE`` or``dde.data.TimePDE``,do not use `batch_size`,and instead use
```
Comment (lululxvi) [inline deepxde/model.py:537]: Add an empty space after `,`

File: deepxde/data/triple.py [hunk 11]
```diff
@@ -71,16 +71,24 @@ def __init__(self, X_train, y_train, X_test, y_test):
         self.train_x, self.train_y = X_train, y_train
         self.test_x, self.test_y = X_test, y_test
 
-        self.train_sampler = BatchSampler(len(X_train[0]), shuffle=True)
+        self.branch_sampler = BatchSampler(len(X_train[0]), shuffle=True)
+        self.trunk_sampler = BatchSampler(len(X_train[1]), shuffle=True)
```
Comment (lululxvi) [inline deepxde/data/triple.py:75]: The comment in line 57 and 58 should also be modified.

File: deepxde/model.py (hunk 8 above)
Comment (Jerry-Jzy) [reply deepxde/model.py:536]: how to preview the docs?

File: deepxde/model.py [hunk 12]
```diff
@@ -533,11 +533,17 @@ def train(
...
-                `dde.callbacks.PDEResidualResampler
-                <https://deepxde.readthedocs.io/en/latest/modules/deepxde.html#deepxde.callbacks.PDEResidualResampler>`_,
-                see an `example <https://github.com/lululxvi/deepxde/blob/master/examples/diffusion_1d_resample.py>`_.
+            batch_size: Integer, tuple, or ``None``.
+
+                - If you solve PDEs via ``dde.data.PDE`` or``dde.data.TimePDE``, do not use `batch_size`,and instead use
+                    `dde.callbacks.PDEResidualResampler
+                    <https://deepxde.readthedocs.io/en/latest/modules/deepxde.html#deepxde.callbacks.PDEResidualResampler>`_,
+                    see an `example <https://github.com/lululxvi/deepxde/blob/master/examples/diffusion_1d_resample.py>`_.
+
```
Comment (lululxvi) [inline deepxde/model.py:542]: Delete the empty line here.

File: deepxde/model.py [hunk 13]
```diff
@@ -533,11 +533,17 @@ def train(
...
                 of times the network weights are updated.
-            batch_size: Integer or ``None``. If you solve PDEs via ``dde.data.PDE`` or
-                ``dde.data.TimePDE``, do not use `batch_size`, and instead use
-                `dde.callbacks.PDEResidualResampler
-                <https://deepxde.readthedocs.io/en/latest/modules/deepxde.html#deepxde.callbacks.PDEResidualResampler>`_,
-                see an `example <https://github.com/lululxvi/deepxde/blob/master/examples/diffusion_1d_resample.py>`_.
+            batch_size: Integer, tuple, or ``None``.
+
+                - If you solve PDEs via ``dde.data.PDE`` or``dde.data.TimePDE``, do not use `batch_size`,and instead use
+                    `dde.callbacks.PDEResidualResampler
```
Comment (lululxvi) [inline deepxde/model.py:539]: No extra indent. Should be aligned with `If`. The same for other lines.

File: deepxde/model.py [hunk 14]
```diff
@@ -533,11 +533,17 @@ def train(
...
             iterations (Integer): Number of iterations to train the model, i.e., number
                 of times the network weights are updated.
-            batch_size: Integer or ``None``. If you solve PDEs via ``dde.data.PDE`` or
-                ``dde.data.TimePDE``, do not use `batch_size`, and instead use
-                `dde.callbacks.PDEResidualResampler
-                <https://deepxde.readthedocs.io/en/latest/modules/deepxde.html#deepxde.callbacks.PDEResidualResampler>`_,
-                see an `example <https://github.com/lululxvi/deepxde/blob/master/examples/diffusion_1d_resample.py>`_.
+            batch_size: Integer, tuple, or ``None``.
+
+                - If you solve PDEs via ``dde.data.PDE`` or``dde.data.TimePDE``, do not use `batch_size`,and instead use
```
Comment (lululxvi) [inline deepxde/model.py:538]: `,and`. Be careful...

File: deepxde/data/triple.py [hunk 15]
```diff
@@ -55,7 +55,7 @@ class TripleCartesianProd(Data):
     Args:
         X_train: A tuple of two NumPy arrays. The first element has the shape (`N1`,
             `dim1`), and the second element has the shape (`N2`, `dim2`). The mini-batch
-            is only applied to `N1`.
+            is applied to `N1` and `N2`.
```
Comment (lululxvi) [inline deepxde/data/triple.py:58]: Delete "The mini-batch is only applied to `N1` and `N2`."

File: deepxde/model.py [hunk 16]
```diff
@@ -533,11 +533,17 @@ def train(
...
+
+                - If you solve PDEs via ``dde.data.PDE`` or``dde.data.TimePDE``, do not use `batch_size`, and instead use
                 `dde.callbacks.PDEResidualResampler
                 <https://deepxde.readthedocs.io/en/latest/modules/deepxde.html#deepxde.callbacks.PDEResidualResampler>`_,
                 see an `example <https://github.com/lululxvi/deepxde/blob/master/examples/diffusion_1d_resample.py>`_.
+
+                - For DeepONet in the format of Cartesian product, if `batch_size` is an Integer,
+                then it is the batch size for the branch input; if you want to also use mini-batch for the trunk net input,
+                set `batch_size` as a tuple, where the fist number is the batch size for the branch net input
+                and the second number is the batch size for the trunk net input.
```
Comment (lululxvi) [inline deepxde/model.py:546]:
```
- If xxxxx
  xxxx
  xxxx
- For DeepONet
  xxx
  xxx
```

Comment (Jerry-Jzy) [reply deepxde/model.py]: fixed

File: deepxde/model.py [hunk 17]
```diff
@@ -533,11 +533,17 @@ def train(
...
-                `dde.callbacks.PDEResidualResampler
-                <https://deepxde.readthedocs.io/en/latest/modules/deepxde.html#deepxde.callbacks.PDEResidualResampler>`_,
-                see an `example <https://github.com/lululxvi/deepxde/blob/master/examples/diffusion_1d_resample.py>`_.
+            batch_size: Integer, tuple, or ``None``.
+
+                - If you solve PDEs via ``dde.data.PDE`` or``dde.data.TimePDE``, do not use `batch_size`, and instead use
+                  `dde.callbacks.PDEResidualResampler
+                  <https://deepxde.readthedocs.io/en/latest/modules/deepxde.html#deepxde.callbacks.PDEResidualResampler>`_,
+                  see an `example <https://github.com/lululxvi/deepxde/blob/master/examples/diffusion_1d_resample.py>`_.
+
```
Comment (lululxvi) [inline deepxde/model.py:542]: Delete this empty line 542.

File: deepxde/data/quadruple.py [hunk 18]
```diff
@@ -48,7 +48,7 @@ class QuadrupleCartesianProd(Data):
     Args:
         X_train: A tuple of three NumPy arrays. The first element has the shape (`N1`,
             `dim1`), the second element has the shape (`N1`, `dim2`), and the third
-            element has the shape (`N2`, `dim3`).
+            element has the shape (`N2`, `dim3`).The mini-batch is applied to `N1` and `N2`.
```
Comment (lululxvi) [inline deepxde/data/quadruple.py:51]: This file should be formatted in the same way as `TripleCartesianProd`

File: deepxde/model.py (hunk 17 above)
Comment (Jerry-Jzy) [reply deepxde/model.py]: fixed

**Pull request merged**